File: desktop/src/main/kotlin/com/fly/desktop/core/network/PeerConnection.kt
package com.fly.desktop.core.network

import com.fly.desktop.core.discovery.NetworkDevice
import com.fly.desktop.core.protocol.FlyMessage
import com.fly.desktop.core.protocol.readFrame
import com.fly.desktop.core.protocol.toFrame
import com.fly.desktop.core.protocol.toMessage
import com.fly.desktop.core.protocol.writeFrame
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.DataInputStream
import java.io.DataOutputStream
import java.net.Socket

class PeerConnection(
    private val socket: Socket,
    val device: NetworkDevice?,
    private val scope: CoroutineScope
) {
    private val output = DataOutputStream(socket.getOutputStream().buffered(65536))
    private val input = DataInputStream(socket.getInputStream().buffered(65536))

    private val _incoming = MutableSharedFlow<Pair<FlyMessage, ByteArray>>(extraBufferCapacity = 128)
    val incoming: SharedFlow<Pair<FlyMessage, ByteArray>> = _incoming

    private val _disconnected = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
    val disconnected: SharedFlow<Unit> = _disconnected

    val isConnected: Boolean get() = socket.isConnected && !socket.isClosed

    fun start() {
        scope.launch(Dispatchers.IO) {
            runCatching {
                while (isConnected) {
                    val frame = readFrame(input)
                    val msg = frame.toMessage() ?: continue
                    _incoming.emit(msg to frame.body)
                }
            }
            _disconnected.emit(Unit)
        }
    }

    suspend fun send(message: FlyMessage, body: ByteArray = ByteArray(0)) = withContext(Dispatchers.IO) {
        runCatching { writeFrame(output, message.toFrame(body)) }
    }

    fun close() = runCatching { socket.close() }
}

File: desktop-cpp/src/network/peer_connection.cpp
#include "peer_connection.h"
#include "../protocol/fly_frame.h"
#include <unistd.h>

PeerConnection::PeerConnection(int fd) : fd_(fd) {}

PeerConnection::~PeerConnection() {
    close();
}

void PeerConnection::start(MsgCallback on_msg, DiscCallback on_disconnect) {
    read_thread_ = std::jthread([this, on_msg, on_disconnect](std::stop_token st) {
        while (!st.stop_requested() && !closed_) {
            try {
                FlyFrame frame = read_frame(fd_);
                auto msg = fly::parse_frame(frame);
                if (msg && on_msg)
                    on_msg(*msg, frame.body);
            } catch (...) {
                break;
            }
        }
        closed_ = true;
        if (on_disconnect) on_disconnect();
    });
}

bool PeerConnection::send(const fly::Message& msg, const std::vector<uint8_t>& body) {
    if (closed_) return false;
    try {
        std::lock_guard<std::mutex> lk(write_mtx_);
        write_frame(fd_, fly::to_frame(msg, body));
        return true;
    } catch (...) {
        return false;
    }
}

void PeerConnection::close() {
    if (!closed_.exchange(true))
        ::close(fd_);
}
